vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+ * The cost is calculated as the number of links in the traffic flow multiplied
+ * by the traffic flow bandwidth. This is then scaled by the priority of the
+ * traffic flow.
+ * 
+ * @param traffic_flow_route The routed path for a traffic flow. This
+ * contains a collection of links in the NoC.
+ * @param traffic_flow_info  Contains the traffic flow bandwidth and
+ * its priority.
+ */
+double calculate_traffic_flow_aggregate_bandwidth_cost(const std::vector<NocLinkId>& traffic_flow_route, const t_noc_traffic_flow& traffic_flow_info);
+
+/**
+ * @brief Determines the latency cost of a routed traffic flow.
+ * The cost is calculated as the combination of the traffic flow
+ * latency and its gap to the traffic flow latency constraint. Each
+ * parameter above is scaled by a weighting factor that determines
+ * the importance each term has on the placement cost. These weightings
+ * are provided by the user. This is then scaled by the priority of the
+ * traffic flow. 
+ * 
+ * 
+ * @param traffic_flow_route The routed path for a traffic flow. This
+ * contains a collection of links in the NoC.
+ * @param noc_model Contains noc information such as the router and link
+ * latencies.
+ * @param traffic_flow_info Contains the traffi flow priority.
+ * @param noc_opts Contains the user provided weightings of the traffic flow 
+ * latency and its constraint parameters for the cost calculation.
+ */
+double calculate_traffic_flow_latency_cost(const std::vector<NocLinkId>& traffic_flow_route, const NocStorage& noc_model, const t_noc_traffic_flow& traffic_flow_info, const t_noc_opts& noc_opts);
+
+/**
+ * @brief Goes through all the traffic flows and determines whether the
+ * latency constraints have been met for each traffic flow. 
+ * 
+ * @return The total number of traffic flows with latency constraints being met
+ */
+int get_number_of_traffic_flows_with_latency_cons_met(void);
+
+/**
+ * @brief THis verifies whether we need to compare the recomputed noc latency
+ * cost to the current noc latency cost. During placement there is a function
+ * called "recompute_costs_from_scratch" that regularily checks whether the
+ * NoC latency cost updates after each move match a complete recalculation of
+ * the NoC latency cost at the current state. This will fail when the latency
+ * costs get really low. So we expect the latency costs to never actually go 
+ * below one picosecond, so we check whether a recomputed cost is below this
+ * threshold and if it is then indicate that there is no need to verify the
+ * NoC latency cost. This is acceptable since the only case where the NoC
+ * latency cost will be below one picosecond is when the weighting on the 
+ * latency component of the cost is extremely low (ie. 0) and when most if not 
+ * all the latency constraints have been met (so this is also close to 0). There
+ * is no need to check and compare the cost here.
+ * 
+ */
+bool check_recomputed_noc_latency_cost(float recomputed_cost);
```

Comment:
Could this code instead be moved into a "check_NoC_costs_equivalent" routine instead?  Makes a somewhat high-level function.

Reply:
Seems like it might be a nicer organization as this helper is probably very short and very specific.